From db42716699e12f7d5d4c91dc5af1edb5e871fa2f Mon Sep 17 00:00:00 2001
From: Min He <min.he@intel.com>
Date: Fri, 7 Sep 2018 06:04:45 +0000
Subject: [PATCH 1629/1676] drm/i915/gvt: Fixed an issue in PS MMIO handlers

When GVT-g guest tries to disable a scaler, it will write 0 to
PS_CTRL register. However, in such case, GVT-g PS MMIO handler will
not allow this value to be written to the HW, which will cause scaler
cannot be disabled.

This patch fixed the issue mentioned above.

V2: Uses PS_SCALER_EN to check if scaler is enabled.

Signed-off-by: Min He <min.he@intel.com>
Signed-off-by: Fei Jiang <fei.jiang@intel.com>
Reviewed-by: Zhao Yakui <yakui.zhao@intel.com>
---
 drivers/gpu/drm/i915/gvt/handlers.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/gvt/handlers.c b/drivers/gpu/drm/i915/gvt/handlers.c
index 676a2d7..d3fc690 100644
--- a/drivers/gpu/drm/i915/gvt/handlers.c
+++ b/drivers/gpu/drm/i915/gvt/handlers.c
@@ -845,9 +845,9 @@ static int skl_ps_mmio_write(struct intel_vgpu *vgpu, unsigned int offset,
 	if (!(vgpu_vreg(vgpu, PIPECONF(pipe)) & I965_PIPECONF_ACTIVE))
 		return 0;
 
-	if (offset == _PS_1A_CTRL || offset == _PS_2A_CTRL ||
+	if ((offset == _PS_1A_CTRL || offset == _PS_2A_CTRL ||
 	   offset == _PS_1B_CTRL || offset == _PS_2B_CTRL ||
-	   offset == _PS_1C_CTRL) {
+	   offset == _PS_1C_CTRL) && ((*(u32 *)p_data) & PS_SCALER_EN)) {
 		unsigned int plane;
 
 		if (SKL_PS_REG_VALUE_TO_PLANE(*(u32 *)p_data) == 0) {
-- 
2.7.4

